### DISCUSSION TITLE
Is there any interest in a GitHub CODEOWNERS file linter that is written in typescript?

### DISCUSSION THREAD

**Discussion by adingler6** [MobileNativeFoundation/discussions#64] [General]
We use [GitHub Code Owners](https://docs.github.com/en/github/creating-cloning-and-archiving-repositories/about-code-owners) in our repositories to enforce code reviews and have created a custom linter to check the CODEOWNERS file for errors on every pull request.

---

Here are some examples of mistakes in the CODEOWNERS file that cause issues:

**Team assigned to a rule does not have WRITE access to the repository**
```
apps/ @octocat # Team octocat does not have WRITE access to the repository
```
In this scenario GitHub gives no indication that something is not working. Instead, files changed in the `apps/` directory will not require code review from anyone.

**Rules are ordered in a way that make certain rules unreachable**
```
source/subdirectory/ @octocat
source/ @source-owner
```
Because the CODEOWNERS file is parsed from the bottom to the top it's impossible for any modified file to trigger the `source/subdirectory/` rule. With very large CODEOWNERS files this can be near impossible to catch without a linter.

**Existence of unescaped spaces**
```
source/cats and dogs/ @octocat
```
This also causes CODEOWNERS to stop working with no indication.

**Special characters cause entire file not work**
```
source/cats&dogs/ @octocat
```
We noticed when rolling out codeowners initially that the `&` symbol caused the entire CODEOWNERS tool to stop working. While it's possible this has been fixed, we added a rule to cover the case.

---

**The linter also has a few other optional checks that can be enabled**
- Verify that every file in the repository is covered by the CODEOWNERS file. This ensures that every new file added in a PR has a matching CODEOWNERS rule
- Verify that all rules in the CODEOWNERS file actually match files on disk. This helps to keep the CODEOWNERS file up to date as files are removed from the codebase

---
### How do we use it in Mobile development?
For iOS and Android we use fastlane, so we `npm install` the linter during the PR pipeline execution:
```
# Simplified fastfile
sh("npm install @codeowners-lint", step_name: "Install Typescript CODEOWNERS linter")
sh("node ./fastlane/node_modules/@codeowners-lint/index.js lint-codeowners")
# Surface errors/warnings as pull request comments
```

**Comment by keith** [comment]
This seems like a nice tool to have. Is it closed source at the moment? Internally at Lyft we have an OWNERS system based on Google's original implementation that predates GitHub's, and we have similar problems like this that we also lint away similarly.

I also kinda think GitHub should try and automate away some of these issues, I'm not sure if a support ticket is the right avenue for that though.

**Reply by adingler6** [reply]
It is currently closed source. And yes, I agree that GitHub should handle all of these checks and surface the errors on the PR that is introducing them or on the repository if it's an issue with team access.
